### ISSUE TITLE
Menu to add other defaultOpen dropdown components, triggers their setOpen event

### ISSUE THREAD

**Issue by cloud-walker** [ariakit/ariakit#2450]
### Current behavior

I think that with the version 2.2 the menu (and maybe other primitives?) started to send an outside click event that triggers the setOpen callback of other popovers.

### Steps to reproduce the bug

1. Open sandbox: https://codesandbox.io/p/sandbox/priceless-gagarin-7tbdfb
2. Interact with the Add select menu
3. adding the fruits select, the select should appear defaultOpen

### Expected behavior

I expect to add a select defaultOpen to remain open without human interaction

### Workaround

https://github.com/ariakit/ariakit/issues/2450#issuecomment-1557403663

### Possible solutions

https://github.com/ariakit/ariakit/issues/2450#issuecomment-1557403663

**Comment by diegohaz** [comment]
Thank you for reporting @cloud-walker!

The focus on hide behavior has been changed in v0.2.2. Now it happens in a microtask. That's likely the reason.

You can temporarily opt into the old behavior using the `autoFocusOnHide` prop:

```jsx
<Ariakit.Menu
  // TODO: Remove when https://github.com/ariakit/ariakit/issues/2450 is fixed
  autoFocusOnHide={(element) => {
    // new behavior
    // queueMicrotask(() => element?.focus());
    // old behavior
    element?.focus();
    // return false to prevent default behavior
    return false;
  }}
/>
```

We do check if there's another element with focus before focusing on hide (which would be the select listbox), but that happens outside the microtask:

https://github.com/ariakit/ariakit/blob/c3aaaa5774a2f9d42fe99a96e893a6a2b98a7944/packages/ariakit-react-core/src/dialog/dialog.tsx#L339

We might need to include the whole logic inside the microtask.

Would you mind sharing your concrete use case for this? It's hard to understand it from the contrived example you provided. Screenshots work. We can create an example and test it to guarantee this won't break again in the future.

**Comment by cloud-walker** [comment]
The use case its a filters bar, with an "Add filter" menu to add a given filter, but the last filter added has not already a value set, so by design we want the picker to be open by default to let the user insert the initial value.

![image](https://github.com/ariakit/ariakit/assets/1144075/c68414c5-95dc-45c9-975f-48748ee07b96)

![image](https://github.com/ariakit/ariakit/assets/1144075/8be5a544-36e7-4052-8e36-59a0c618e76f)

![image](https://github.com/ariakit/ariakit/assets/1144075/2633dbb2-afe2-4bf2-9684-c169cb7c54ff)

**Comment by cloud-walker** [comment]
The picker with defaultOpen is the closed automatically if the open state changes and no value is provided:


<img width="519" alt="image" src="https://github.com/ariakit/ariakit/assets/1144075/b702c660-0f13-417b-926c-c36cd889ab67">

**Comment by diegohaz** [comment]
I see. So what you were seeing after upgrading to v0.2.3 was no filter select being rendered after adding a filter?

**Comment by cloud-walker** [comment]
Yes I see.. no filter added on selecting it from the menu, (because of the setOpen logic that I've linked), and after investigating a bit I discovered that the setOpen is called automatically after the first render, causing (in my case) the filter picker to "auto remove" itself 😄 

as a bonus, the aria-expanded attribute seems not to be set when defaultOpen is true, I don't know if it is a bug or I'm not using the ariakit API right

**Comment by diegohaz** [comment]
> as a bonus, the aria-expanded attribute seems not to be set when defaultOpen is true, I don't know if it is a bug or I'm not using the ariakit API right

I can't reproduce it with the plain select example. I can see it in your CodeSandbox with the workaround I proposed though. But I can also reproduce it with older versions. Was that working before?

You can force the `aria-expanded` prop yourself:

```jsx
const expanded = store.useState("open");
<Select aria-expanded={expanded} />
```

The `aria-expanded` attribute is sort of dynamic because a disclosure content (Dialog, Popover, SelectPopover, Menu, etc.) may have more than one disclosure button:

https://github.com/ariakit/ariakit/blob/c3aaaa5774a2f9d42fe99a96e893a6a2b98a7944/packages/ariakit-react-core/src/disclosure/disclosure.ts#L34-L45

The active disclosure button (which will ultimately get the `aria-expanded` attribute) is also determined by the currently focused element right before a dialog is open:

https://github.com/ariakit/ariakit/blob/c3aaaa5774a2f9d42fe99a96e893a6a2b98a7944/packages/ariakit-react-core/src/disclosure/disclosure.ts#L34-L45

I'm not sure what exactly happened there, but I think that's one of the reasons we moved the autofocus-on-hide behavior to a microtask.

**Comment by diegohaz** [comment]
The fix should be available in v0.2.4

I created an example based on your use case here: https://ariakit.org/examples/select-menu-default-open
